flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointPlanCalculator.java

```diff
@@ -18,109 +18,498 @@
-                ackTasks.put(ee.getAttemptId(), ev);
+    /**
+     * Compute the accurate running tasks for each job vertex. Currently if multiple tasks all
+     * finished in short period, the order of their reports of FINISHED is nondeterministic, and
+     * some tasks may report FINISHED before all its precedent tasks have.
+     *
+     * <p>To overcome this issue we would iterates the graph first to acquire the accurate running
+     * tasks. We would iterate the job graph in reverse topological order, and for each job vertex,
+     * we would remove those precedent tasks that connected to finished tasks of this job vertex
+     * from possibly running tasks.
+     *
+     * @return An accurate set of running tasks for each job vertex.
+     */
+    @VisibleForTesting
+    Map<JobVertexID, JobVertexTaskSet> calculateRunningTasks() {
+        Map<JobVertexID, JobVertexTaskSet> runningTasksByVertex = new HashMap<>();
+
+        ListIterator<ExecutionJobVertex> jobVertexIterator =
+                jobVerticesInTopologyOrder.listIterator(jobVerticesInTopologyOrder.size());
+
+        while (jobVertexIterator.hasPrevious()) {
+            ExecutionJobVertex jobVertex = jobVertexIterator.previous();
+
+            List<JobEdge> outputJobEdges = getOutputJobEdges(jobVertex);
+
+            // we're lucky if this is true
+            if (isFinishedAccordingToDescendants(runningTasksByVertex, outputJobEdges)) {
+                runningTasksByVertex.put(
+                        jobVertex.getJobVertexId(), JobVertexTaskSet.noTasks(jobVertex));
+                continue;
+            }
+
+            // not lucky, need to determine which of our tasks can still be running
+            Set<ExecutionVertexID> runningTasks =
+                    getRunningTasks(runningTasksByVertex, jobVertex, outputJobEdges);
+
+            runningTasksByVertex.put(
+                    jobVertex.getJobVertexId(),
+                    JobVertexTaskSet.someTasks(jobVertex, runningTasks));
+        }
+
+        return runningTasksByVertex;
+    }
+
+    /**
+     * Determines the {@link ExecutionVertexID ExecutionVertexIDs} of those subtasks that are still
+     * running.
+     */
+    private Set<ExecutionVertexID> getRunningTasks(
+            Map<JobVertexID, JobVertexTaskSet> runningTasksByVertex,
+            ExecutionJobVertex jobVertex,
+            List<JobEdge> outputJobEdges) {
+        return Arrays.stream(jobVertex.getTaskVertices())
+                .filter(
+                        task -> {
+                            // the task is done if the execution graph thinks it is done
+                            if (task.getCurrentExecutionAttempt().isFinished()) {
+                                return false;
+                            }
+
+                            for (JobEdge edge : outputJobEdges) {
+                                if (edge.getDistributionPattern()
+                                        == DistributionPattern.POINTWISE) {
```

Comment:
Why do we check this only for `POINTWISE` connections? For `ALL_TO_ALL` downstream subtask to stop, all its upstreams must send `EndOfPartition`, no?

Or is it an optimization to avoid O(n^2)?